serialization/api.go

```diff
@@ -214,72 +214,101 @@ type PositionalDataOutput interface {
 }
 
 // DataInput provides deserialization methods.
+// If any of the methods results in an error, all following methods will return the zero value
+// for that type immediately.
+// Example usage:
+//  field1 = input.ReadUTF()
+//  field2 = input.ReadUTF()
+//  return input.Error()
 type DataInput interface {
+	// Error returns the first error encountered by DataInput.
+	Error() error
+
 	// Position returns the head position in the byte array.
 	Position() int32
 
 	// SetPosition sets the head position in the byte array.
 	SetPosition(pos int32)
 
-	// ReadByte returns byte read and error.
-	ReadByte() (byte, error)
+	// ReadByte returns byte read .
+	// It returns the zero value if an error is set previously.
+	ReadByte() byte
 
-	// ReadBool returns bool read and error.
-	ReadBool() (bool, error)
+	// ReadBool returns bool read.
+	// It returns the zero value if an error is set previously.
+	ReadBool() bool
 
-	// ReadUInt16 returns uint16 read and error.
-	ReadUInt16() (uint16, error)
+	// ReadUInt16 returns uint16 read.
+	// It returns the zero value if an error is set previously.
+	ReadUInt16() uint16
 
-	// ReadInt16 returns int16 read and error.
-	ReadInt16() (int16, error)
+	// ReadInt16 returns int16 read.
+	// It returns the zero value if an error is set previously.
+	ReadInt16() int16
 
-	// ReadInt32 returns int32 read and error.
-	ReadInt32() (int32, error)
+	// ReadInt32 returns int32 read.
+	// It returns the zero value if an error is set previously.
+	ReadInt32() int32
 
-	// ReadInt64 returns int64 read and error.
-	ReadInt64() (int64, error)
+	// ReadInt64 returns int64 read.
+	// It returns the zero value if an error is set previously.
+	ReadInt64() int64
 
-	// ReadFloat32 returns float32 read and error.
-	ReadFloat32() (float32, error)
+	// ReadFloat32 returns float32 read.
+	// It returns the zero value if an error is set previously.
+	ReadFloat32() float32
 
-	// ReadFloat64 returns float64 read and error.
-	ReadFloat64() (float64, error)
+	// ReadFloat64 returns float64 read.
+	// It returns the zero value if an error is set previously.
+	ReadFloat64() float64
 
-	// ReadUTF returns string read and error.
-	ReadUTF() (string, error)
+	// ReadUTF returns string read.
+	// It returns the zero value if an error is set previously.
+	ReadUTF() string
 
-	// ReadObject returns object read and error.
-	ReadObject() (interface{}, error)
+	// ReadObject returns object read.
+	// It returns the zero value if an error is set previously.
```

Comment:
returns `nil` if an error is set previously. And there are similar problems in other methods. Can recheck them?

Reply:
I used `the zero value` as in a general phrase. For example for `int` it is 0, for `string` it is "", for object it is `nil`, do you think it is unclear?

Reply:
Yes, I think we need to be precise.